### PR TITLE
Fix test_eager_matches_sdpa_inference for XPU backend

### LINKED ISSUE #34888: xpu: test_eager_matches_sdpa_inference tests fail with pytorch XPU backend
With:
* https://github.com/huggingface/transformers/commit/54be2d7ae87e873482b984cc956e165ca4dc0ba3
* https://github.com/huggingface/accelerate/commit/e11d3ceff3a49378796cdff5b466586d877d5c60
* https://github.com/pytorch/pytorch/commit/e429a3b72e787ddcc26ee2ba177643c9177bab24

```
$ cat spec.py
import torch
DEVICE_NAME = 'xpu'
MANUAL_SEED_FN = torch.xpu.manual_seed
EMPTY_CACHE_FN = torch.xpu.empty_cache
DEVICE_COUNT_FN = torch.xpu.device_count

$ TRANSFORMERS_TEST_DEVICE_SPEC=spec.py python3 -m pytest --pspec tests/models -k test_eager_matches_sdpa_inference
<...>
FAILED tests/models/audio_spectrogram_transformer/test_modeling_audio_spectrogram_transformer.py::
    Here we also overwrite some of the tests of test_modeling_common.py, as AST does not use input_ids, inputs_embeds,
    attention_mask and seq_length.
    ::test_eager_matches_sdpa_inference_0_float16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 4.739e-05,...
FAILED tests/models/audio_spectrogram_transformer/test_modeling_audio_spectrogram_transformer.py::
    Here we also overwrite some of the tests of test_modeling_common.py, as AST does not use input_ids, inputs_embeds,
    attention_mask and seq_length.
    ::test_eager_matches_sdpa_inference_1_bfloat16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 5.913e-04,...
FAILED tests/models/bart/test_modeling_bart.py::BartModelTest::test_eager_matches_sdpa_inference_0_float16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 7.510e-06,...
FAILED tests/models/bart/test_modeling_bart.py::BartModelTest::test_eager_matches_sdpa_inference_1_bfloat16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 7.772e-05,...
FAILED tests/models/bart/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_eager_matches_sdpa_inference_0_float16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 2.402e-05,...
FAILED tests/models/bart/test_modeling_bart.py::BartStandaloneDecoderModelTest::test_eager_matches_sdpa_inference_1_bfloat16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 3.490e-04,...
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_eager_matches_sdpa_inference_0_float16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 5.555e-05,...
FAILED tests/models/bert/test_modeling_bert.py::BertModelTest::test_eager_matches_sdpa_inference_1_bfloat16 - AssertionError: False is not true : padding_side=left, use_mask=False, enable_kernels=False: mean relative difference: 3.567e-04,...
<...>
======================= 159 failed, 89 passed, 793 skipped, 75366 deselected, 319 warnings in 74.89s (0:01:14) =======================
```

CC: @amyeroberts @ydshieh

### PR DESCRIPTION
**Pull request by dvrogozh**
Included fixes:
* Use `torch.nn.attention.sdpa_kernel` instead of deprecated `torch.backends.cuda.sdp_kernel`
* Use `torch.amp.autocast` instead of deprecated `torch.cuda.amp.autocast` in nemotron
* Reuse CUDA MATH thresholds in for XPU (as of PyTorch 2.5 XPU backend supports only `torch.nn.attention.SDPBackend.MATH`)

Fixes: #34888
CC: @amyeroberts @ydshieh

### CODE REVIEW

**Comment by dvrogozh** [comment]
Looks like Friday evening is not the best time to run ci. Pushed same code 3 times, seeing different errors on each run:). Not related to the change I think. Will continue on Monday :).

**Comment by dvrogozh** [comment]
Clarified XPU backend behavior for `torch.backends.cuda.sdp_kernel`. As of PyTorch 2.5 (and today PyTorch main), XPU backend supports only `torch.nn.attention.SDPBackend.MATH` implementation of which is device agnostic with respect to implementation of each individual aten operator. So, we can reuse CUDA or CPU MATH weights for XPU (for all cases in disrespect to SDP backend). That's the change I did in the last version.

File: tests/models/musicgen/test_modeling_musicgen.py
Comment (faaany) [inline tests/models/musicgen/test_modeling_musicgen.py:636]: why `enable_kernels` not set to False here?

File: tests/models/musicgen_melody/test_modeling_musicgen_melody.py
Comment (faaany) [inline tests/models/musicgen_melody/test_modeling_musicgen_melody.py:644]: the same here.

File: tests/models/musicgen/test_modeling_musicgen.py
Comment (dvrogozh) [reply tests/models/musicgen/test_modeling_musicgen.py]: I missed it. Fixed.

File: tests/models/musicgen_melody/test_modeling_musicgen_melody.py
Comment (dvrogozh) [reply tests/models/musicgen_melody/test_modeling_musicgen_melody.py]: fixed

File: tests/models/mimi/test_modeling_mimi.py
Comment (ydshieh) [inline tests/models/mimi/test_modeling_mimi.py:660]: nit:

although xpu only support MATH, does it means the results from XPU will be the same as, say CUDA? Otherwise, I don't see the reason to reuse CUDA threshold.

**Comment by ydshieh** [comment]
Thanks @dvrogozh ! LGTM.

Have you run (some of) the relevant tests on a GPU (cuda) and XPU (cc @faaany  for this part maybe) machine?

File: tests/models/mimi/test_modeling_mimi.py
Comment (dvrogozh) [reply tests/models/mimi/test_modeling_mimi.py:660]: In my understanding, at least for the recent versions of PyTorch (2.5 and upcoming 2.6 ) MATH should give identical results on any hardware including different GPU devices and CPU because algorithm is implemented on torch level and is device agnostic (well, up to aten operators implementation which is device specific, but they still should give same results). The only exception here is MPS which has separate branch in the code though also implemented at torch level. Here are relevant places in the sources:
* https://github.com/pytorch/pytorch/blob/5ececd4caa4ec1534c567ec9db6efb861b760685/aten/src/ATen/native/transformers/attention.cpp#L776 (see separate code branch for MPS right above this code)
* https://github.com/pytorch/pytorch/blob/5ececd4caa4ec1534c567ec9db6efb861b760685/aten/src/ATen/native/transformers/attention.cpp#L795

Comment (ydshieh) [reply tests/models/mimi/test_modeling_mimi.py]: sound reason able to reuse (for MATH) threshold now. Thank you for explaining!

**Review by ydshieh — APPROVED** [review]
Thank you again! 

@faaany Do you have any further comments?

**Comment by dvrogozh** [comment]
For myself: need to follow up with #34941 (sdpa tests for beit) which is being reviewed in parallel.

**Comment by dvrogozh** [comment]
```
FAILED tests/models/xlm/test_modeling_xlm.py::XLMModelTest::test_batching_equivalence - AssertionError: tensor(False) is not true : Batched and Single row outputs are not equal in XLMForQuestionAnswering for key=end_top_index. Difference=1.
```
Failure on ci seems unrelated. I also see some flakiness on main ci results.

**Comment by ydshieh** [comment]
yes we do have some flaky tests. (trying to fix them but in a slow peace)

**Review by ruidazeng — APPROVED** [review]

**Comment by ydshieh** [comment]
There is no need to have a green CI if you think some failure are irrelevant. Ping me for a double check then we could wait for a core maintainer's review.

**Comment by dvrogozh** [comment]
> There is no need to have a green CI if you think some failure are irrelevant. Ping me for a double check then we could wait for a core maintainer's review.

Yeah, I think that's the case here. I rebased today w/o any changes. Another test failed this time, I believe unrelated as well. Sorry, I am just paranoid about green ci - I consider that's my responsibility to achieve that on my PRs.

```
FAILED tests/trainer/test_trainer.py::TrainerIntegrationWithHubTester::test_push_to_hub_tags - KeyError: 'url'
FAILED tests/trainer/test_trainer.py::TrainerIntegrationWithHubTester::test_push_to_hub_with_saves_each_epoch - AssertionError: no logs of level WARNING or higher triggered on root
```

**Review by ArthurZucker — APPROVED** [review]
Feel free to merge if it's alright with you @ydshieh

**Pull request merged**
